src/interpreter/mod.rs

```diff
@@ -33,18 +36,109 @@ mod error;
 mod inner;
 mod stack;
 
+use crate::MiniscriptKey;
+
 pub use self::error::Error;
+use self::error::PkEvalErrInner;
 use self::stack::Stack;
 
 /// An iterable Miniscript-structured representation of the spending of a coin
 pub struct Interpreter<'txin> {
     inner: inner::Inner,
     stack: Stack<'txin>,
-    script_code: bitcoin::Script,
+    script_code: Option<bitcoin::Script>, // taproot transactions don't have script code
     age: u32,
     height: u32,
 }
 
+// A type representing functions for checking signatures that accept both
+// Ecdsa and Schnorr signatures
+
+/// A type for representing signatures supported as of bitcoin core 22.0
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum KeySigPair {
+    /// A Full public key and corresponding Ecdsa signature
+    Ecdsa(bitcoin::PublicKey, bitcoin::EcdsaSig),
+    /// A x-only key and corresponding Schnorr signature
+    Schnorr(bitcoin::XOnlyPublicKey, bitcoin::SchnorrSig),
+}
+
+// Internally used enum for different types of bitcoin keys
+// Even though we implement MiniscriptKey for BitcoinKey, we make sure that there
+// are little mis-use
+// - The only constructors for this are only called in from_txdata that take care
+//   using the correct enum variant
+// - This does not implement ToPublicKey to avoid context dependant encoding/decoding of 33/32
+//   byte keys. This allows us to keep a single NoChecks context instead of a context for
+//   for NoChecksSchnorr/NoChecksEcdsa.
+// Long term TODO: There really should be any need for Miniscript<Pk: MiniscriptKey> struct
+// to have the Pk: MiniscriptKey bound. The bound should be on all of it's methods. That would
+// require changing Miniscript struct to three generics Miniscript<Pk, Pkh, Ctx> and bound on
+// all of the methods of Miniscript to ensure that Pkh = Pk::Hash
+#[derive(Hash, Eq, Ord, PartialEq, PartialOrd, Clone, Copy, Debug)]
+enum BitcoinKey {
+    // Full key
+    Fullkey(bitcoin::PublicKey),
+    // Xonly key
+    XOnlyPublicKey(bitcoin::XOnlyPublicKey),
+}
+
+// Displayed in full 33 byte representation. X-only keys are displayed with 0x02 prefix
+impl fmt::Display for BitcoinKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            BitcoinKey::Fullkey(pk) => pk.to_public_key().fmt(f),
+            BitcoinKey::XOnlyPublicKey(pk) => pk.to_public_key().fmt(f),
+        }
+    }
+}
+
+impl From<bitcoin::PublicKey> for BitcoinKey {
+    fn from(pk: bitcoin::PublicKey) -> Self {
+        BitcoinKey::Fullkey(pk)
+    }
+}
+
+impl From<bitcoin::XOnlyPublicKey> for BitcoinKey {
+    fn from(xpk: bitcoin::XOnlyPublicKey) -> Self {
+        BitcoinKey::XOnlyPublicKey(xpk)
+    }
+}
+
+// While parsing we need to remember how to the hash was parsed so that we can
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+enum TaggedHash160 {
```

Comment:
what does "tagged" mean in this context?

just that we track what kind of key?

May call it TypedHash because TaggedHash now means something else